xls/scheduling/pipeline_schedule.cc

```diff
@@ -141,6 +144,175 @@ absl::StatusOr<ScheduleCycleMap> ScheduleToMinimizeRegisters(
   return cycle_map;
 }
 
+// A helper function to compute each node's delay by calling the delay estimator
+// The result is used by `ComputeCriticalCombinationalPaths`,
+// `ComputeSingleSourceCCP`, `ScheduleToMinimizeRegistersSDC`
+absl::StatusOr<absl::flat_hash_map<Node*, int64_t>> ComputeNodeDelay(
+    FunctionBase* f, const DelayEstimator& delay_estimator) {
+  absl::flat_hash_map<Node*, int64_t> result;
+  for (Node* node : f->nodes()) {
+    XLS_ASSIGN_OR_RETURN(result[node],
+                         delay_estimator.GetOperationDelayInPs(node));
+  }
+  return result;
+}
+
+// A helper function to compute single source Critical Combinational Paths'
+// delay, from the src-node to all other reachable nodes.
+//
+// It traverses the data-dependence graph from the src-node,
+// compute and record the Critical Combinational Path's delay.
+//
+// Args:
+//   src: the starting node
+//   node_delay: node's delay, which is precomputed from delay estimator
+absl::flat_hash_map<Node*, int64_t> ComputeSingleSourceCCP(
+    Node* src, const absl::flat_hash_map<Node*, int64_t>& node_delay) {
+  absl::flat_hash_map<Node*, int64_t> dst_ccp;
+  absl::flat_hash_set<Node*> worklist;
+
+  auto pop_worklist = [&worklist]() {
+    auto it = worklist.begin();
+    Node* result = *it;
+    worklist.erase(it);
+    return result;
+  };
+
+  // Topological Sort.
+  worklist.insert(src);
+  while (worklist.size() > 0) {
+    Node* node = pop_worklist();
+
+    for (Node* user : node->users()) {
+      dst_ccp[user] =
+          std::max(dst_ccp[user], dst_ccp[node] + node_delay.at(node));
+
+      worklist.insert(user);
+    }
+  }
+
+  return dst_ccp;
+}
+
+// Compute Critical Combinational Path's delay between any pair of nodes.
+//
+// Complexity: O( |V|·(|V| + |E|) )
+//
+// The result maps path's (src, dst) pair to the path's combinational delay.
+std::vector<std::pair<std::pair<Node*, Node*>, int64_t>>
+ComputeCriticalCombinationalPaths(
+    FunctionBase* f, const absl::flat_hash_map<Node*, int64_t>& node_delay) {
+  using PathTy =
+      std::pair<Node*, Node*>;  // describe a path, from src-node to dst-node
+  std::vector<std::pair<PathTy, int64_t>>
+      result;  // critical combinational path's delay
+
+  for (Node* src : f->nodes()) {
+    for (auto [dst, ccp] : ComputeSingleSourceCCP(src, node_delay)) {
+      std::pair<Node*, Node*> path(src, dst);
+      result.push_back(std::make_pair(path, ccp));
+    }
+  }
+  return result;
+}
+
+using namespace operations_research;
```

Comment:
`using namespace` is against our style guide. I think `namespace or_tools = ::operations_research;` would be acceptable.